src/librustc/ty/wf.rs

```diff
@@ -157,33 +162,77 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
                 .collect()
     }
 
-    /// Pushes the obligations required for `trait_ref` to be WF into
-    /// `self.out`.
+    /// Pushes the obligations required for `trait_ref` to be WF into `self.out`.
     fn compute_trait_ref(&mut self, trait_ref: &ty::TraitRef<'tcx>, elaborate: Elaborate) {
+        let tcx = self.infcx.tcx;
         let obligations = self.nominal_obligations(trait_ref.def_id, trait_ref.substs);
 
         let cause = self.cause(traits::MiscObligation);
         let param_env = self.param_env;
 
         if let Elaborate::All = elaborate {
+            let trait_assoc_items = tcx.associated_items(trait_ref.def_id);
+
             let predicates = obligations.iter()
-                                        .map(|obligation| obligation.predicate.clone())
-                                        .collect();
-            let implied_obligations = traits::elaborate_predicates(self.infcx.tcx, predicates);
+                .map(|obligation| obligation.predicate.clone())
+                .collect();
+            let implied_obligations = traits::elaborate_predicates(tcx, predicates);
+            let item_span: Option<Span> = self.item.map(|i| i.span);
+            let item = &self.item;
             let implied_obligations = implied_obligations.map(|pred| {
-                traits::Obligation::new(cause.clone(), param_env, pred)
+                let mut cause = cause.clone();
+                match &pred {
```

Comment:
this is kind of a big blob of complex code -- can we pull it into a suggestively named helper function, with a comment? If I'm not mistaken, it basically makes a "best effort" to compute a better cause, right? I'd like to see this be something like:

```rust
let cause = compute_cause_for_elaborated_predicate(...);
traits::Obligation::new(cause, param_env, pred)
```

ideally, with some comments in the fn body explaining what is going on

Reply:
I moved it to a closure to avoid multiple mutable borrowing of `self`, which is less than ideal but should help with following what's going on.

Reply:
Done.